### PR TITLE
Clean up anomalous uses of tmp registers

### LINKED ISSUE #941: Clean up anomalous uses of tmp registers
Currently we assign to tempRegisters "internally" in the following places

1. https://github.com/ballerina-platform/nballerina/blob/0005efda65f580233f63a8485cd5ad774e5feb4d/compiler/modules/front/stmt.bal#L1161
2. https://github.com/ballerina-platform/nballerina/blob/0005efda65f580233f63a8485cd5ad774e5feb4d/compiler/modules/front/expr.bal#L312
3. https://github.com/ballerina-platform/nballerina/blob/0005efda65f580233f63a8485cd5ad774e5feb4d/compiler/modules/front/expr.bal#L317
4. https://github.com/ballerina-platform/nballerina/blob/0005efda65f580233f63a8485cd5ad774e5feb4d/compiler/modules/front/expr.bal#L696
5. https://github.com/ballerina-platform/nballerina/blob/0005efda65f580233f63a8485cd5ad774e5feb4d/compiler/modules/front/expr.bal#L698

_Originally posted by @heshanpadmasiri in https://github.com/ballerina-platform/nballerina/pull/939#discussion_r811546704_

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Clean up anomalous uses of tmp registers
Resolves to #941

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/front/stmt.bal [hunk 1]
```diff
@@ -1158,7 +1158,17 @@ function codeGenExprForCond(StmtContext cx, bir:BasicBlock bb, Environment env,
     }
     else if flags != 0 {
         bir:Register reg = cx.createTmpRegister(t:BOOLEAN);
-        bir:AssignInsn insn = { result: reg, operand: { value, semType: t:BOOLEAN }, pos: expr.startPos };
+        bir:EqualityOp op;
+        bir:Operand equalityOperand;
+        if value {
+            equalityOperand = { value: false, semType: t:BOOLEAN };
```
Comment (jclark) [inline compiler/modules/front/stmt.bal]: I don't understand what you are doing here. Why does it need to be false here and true in the other branch?

Shouldn't this use t:singleton?

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/stmt.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/stmt.bal]: I think I misunderstood what you said in the [comment](https://github.com/ballerina-platform/nballerina/issues/941#issuecomment-1048373596). Fixed

File: compiler/modules/front/stmt.bal [hunk 2]
```diff
@@ -1158,7 +1158,12 @@ function codeGenExprForCond(StmtContext cx, bir:BasicBlock bb, Environment env,
     }
     else if flags != 0 {
         bir:Register reg = cx.createTmpRegister(t:BOOLEAN);
-        bir:AssignInsn insn = { result: reg, operand: { value, semType: t:BOOLEAN }, pos: expr.startPos };
+        bir:EqualityInsn insn = {
+            op: "===",
```
Comment (jclark) [inline compiler/modules/front/stmt.bal]: This is violating the rule that the intersection of the types of the operands of === must not be disjoint.

You need to use `===` and `!==` ops with the equal operands, as I suggested.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/stmt.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/stmt.bal]: fixed

**Comment by heshanpadmasiri** [comment]
Replaced `TmpRegister` with `AssignTmpRegister` in `AssignInsn`

File: compiler/modules/front/stmt.bal [hunk 3]
```diff
@@ -911,7 +911,7 @@ function lookupVarRefForAssign(StmtContext cx, Environment env, string varName,
 
 function codeGenAssign(StmtContext cx, Environment env, bir:BasicBlock block, bir:Register result, s:Expr expr, t:SemType semType, Position pos) returns CodeGenError|bir:BasicBlock {
     var { result: operand, block: nextBlock } = check cx.codeGenExpr(block, env, semType, expr);
-    if result !is bir:TmpRegister|bir:VarRegister|bir:FinalRegister {
+    if result !is bir:AssignTmpRegister|bir:VarRegister|bir:FinalRegister {
```
Comment (jclark) [inline compiler/modules/front/stmt.bal]: We should be passing in the right kind of register here not casting. This means more changes are still needed. I guess Binding needs changing to use the right kind of register.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/stmt.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/stmt.bal]: We currently use bindings with register to which you can't assign values at 2 places

1. https://github.com/ballerina-platform/nballerina/blob/1db49a0f695ac02032f4e8556e80f212dd6ae89e/compiler/modules/front/stmt.bal#L226
2. https://github.com/ballerina-platform/nballerina/blob/1db49a0f695ac02032f4e8556e80f212dd6ae89e/compiler/modules/front/stmt.bal#L801

If we are to keep allowing this we can also check for register type in [`lookupVarRefForAssign`](https://github.com/ballerina-platform/nballerina/blob/1db49a0f695ac02032f4e8556e80f212dd6ae89e/compiler/modules/front/stmt.bal#L890) and return the correct type (or panic if it is not the correct type)

Comment (jclark) [reply compiler/modules/front/stmt.bal]: The binding will need to allow for
- params
- final variables
- narrowed variables
- normal variables

Only the last one can be assigned to.

The first two correspond to errors in the source: so these cases should be rejected during the lookup as a semantic error. In the third case, the assignment will cause the variable not be narrowed.  There will be a different type for the original binding, which won't allowed for narrowed registers.

There's no need to explicitly check and panic: just do the cast and put a comment explaining why the cast must succeed.

Comment (heshanpadmasiri) [reply compiler/modules/front/stmt.bal]: fixed

File: compiler/modules/bir/bir.bal [hunk 4]
```diff
@@ -165,8 +167,15 @@ public type TmpRegister readonly & record {|
...
 
+public type AssignTmpRegister readonly & record {|
+    *RegisterBase;
+    ASSIGN_TMP_REGISTER_KIND kind;
+|};
+
 public type NarrowRegister readonly & record {|
     *RegisterBase;
+    // position of the register that was narrowed
+    Position prevPos;
```
Comment (jclark) [inline compiler/modules/bir/bir.bal:178]: We need the number of the register not its position.

**Pull request merged**